### PR TITLE
Feature/billing fixes

### PR DESCRIPTION
**Pull request by tuj**
https://jira.itkdev.dk/browse/JE-217

* Added filter to invoice overview.
* Fixed number input type for invoice entries.
* Fixed bug with adding additional worklogs to existing invoice entry.
* Removed option to edit recorded invoice.
* Added sum of invoice entry totals to bottom of invoice table.
* Fixed bug where you could select worklogs or expenses added to another invoice with the "select all" checkbox.
* Removed ApiPlatform since it was not used.

### CODE REVIEW

File: bundles/Billing/Resources/assets/pages/InvoiceEntry.js [hunk 1]
```diff
@@ -432,10 +446,10 @@ export class InvoiceEntry extends Component {
                                 type="text"
                                 name={'totalPrice'}
                                 className="form-control"
-                                id="invoice-entry-unit-price"
+                                id="invoice-entry-total-price"
                                 aria-describedby="enterUnitPrice"
                                 disabled={true}
-                                defaultValue={ this.state.price * this.state.amount }>
+                                value={ this.state.price * this.state.amount }>
```
[the remark this answers is not in the document]
Comment (tuj) [reply bundles/Billing/Resources/assets/pages/InvoiceEntry.js]: False positive.

File: bundles/Billing/Resources/assets/pages/HomePage.js [hunk 2]
```diff
@@ -228,11 +271,27 @@ class HomePage extends Component {
...
+                                        value={creationDateSortingOptions.filter(item => this.state.filterValues.creationDateSorting === item.value)}
+                                        name={'creationDateSorting'}
+                                        aria-label={t('home_page.filter.creator')}
+                                        onChange={(selectedOption) => this.handleFilterChange('creationDateSorting', selectedOption ? selectedOption.value : '')}
+                                        options={creationDateSortingOptions}
+                                    />
+                                    <label htmlFor={'creatorFilter'}>{t('home_page.filter.creator')}</label>
+                                    <Select
+                                        id={'creatorFilter'}
+                                        value={creatorFilterOptions.filter(item => this.state.filterValues.creatorFilter === item.value)}
```
[the remark this answers is not in the document]
Comment (tuj) [reply bundles/Billing/Resources/assets/pages/HomePage.js:286]: False positive.

File: bundles/Billing/Resources/assets/pages/HomePage.js [hunk 3]
```diff
@@ -228,11 +271,27 @@ class HomePage extends Component {
...
                                 <Form.Group className="mb-0">
-                                    <Form.Label className="sr-only">{t('home_page.sort')}</Form.Label>
-                                    <Form.Control size="sm" as="select" value={this.state.sortOrder} onChange={this.toggleSort.bind(this) }>
-                                        <option value={'desc'}>{t('home_page.sorting.newest')}</option>
-                                        <option value={'asc'}>{t('home_page.sorting.oldest')}</option>
-                                    </Form.Control>
+                                    <label htmlFor={'creationDateSorting'}>{t('home_page.sort.created_at')}</label>
+                                    <Select
+                                        id={'creationDateSorting'}
+                                        value={creationDateSortingOptions.filter(item => this.state.filterValues.creationDateSorting === item.value)}
```
[the remark this answers is not in the document]
Comment (tuj) [reply bundles/Billing/Resources/assets/pages/HomePage.js:277]: False positive.

File: bundles/Billing/Resources/assets/pages/InvoiceEntry.js [hunk 4]
```diff
@@ -378,64 +392,66 @@ export class InvoiceEntry extends Component {
...
                             <input
                                 type="text"
+                                id={'totalPrice'}
                                 name={'totalPrice'}
                                 className="form-control"
-                                id="invoice-entry-unit-price"
                                 aria-describedby="enterUnitPrice"
                                 disabled={true}
-                                defaultValue={ this.state.price * this.state.amount }>
+                                value={ this.state.price * this.state.amount }>
```
[the remark this answers is not in the document]
Comment (tuj) [reply bundles/Billing/Resources/assets/pages/InvoiceEntry.js:454]: False positive.

File: bundles/Billing/Resources/assets/pages/InvoiceEntry.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tuj) [reply bundles/Billing/Resources/assets/pages/InvoiceEntry.js:454]: False positive.

File: .env [hunk 5]
```diff
@@ -47,11 +47,6 @@ JIRA_DEFAULT_BOARD=65
 # @see https://confluence.atlassian.com/cloud/api-tokens-938839638.html
 JIRA_USER=""
 JIRA_PASS=""
-
```
Comment (rimi-itk) [inline .env:50]: Why is this removed?

File: bundles/Billing/Service/BillingService.php [hunk 6]
```diff
@@ -30,27 +31,30 @@ class BillingService extends JiraService
...
      * @param $boundCustomFieldMappings
      */
     public function __construct(
-        EntityManagerInterface $entityManager,
         $jiraUrl,
         $tokenStorage,
         $customerKey,
         $pemPath,
+        CacheProvider $cache,
+        EntityManagerInterface $entityManager,
```
Comment (rimi-itk) [inline bundles/Billing/Service/BillingService.php:57]: Is this entity manager actually used? No instance variable is set in the constructor.

File: bundles/Billing/Service/BillingService.php [hunk 7]
```diff
@@ -245,6 +250,10 @@ public function putInvoice($invoiceData)
             throw new HttpException(404, 'Unable to update invoice with id '.$invoiceData['id'].' as it does not already exist');
         }
 
+        if ($invoice->getRecorded()) {
```
Comment (rimi-itk) [inline bundles/Billing/Service/BillingService.php:253]: Just out of curiosity: What does it mean that an invoice has been recorded?

File: src/Entity/AbstractEntity.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        return (string) $this->id;
+    }
+}
```
Comment (rimi-itk) [inline src/Entity/AbstractEntity.php]: Well, this seems like lazy programmer's solution … `static::class . $this-id` may be a better value, but I think we should let the actual entities implement `__toString`.

File: .env (hunk 5 above)
Comment (rimi-itk) [reply .env:50]: Oh, `api-platform` is removed.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: bundles/Billing/Service/BillingService.php (hunk 6 above)
Comment (tuj) [reply bundles/Billing/Service/BillingService.php:57]: Øhh... It is used in multiple places :)
The reason it is in included in the PR, is because I changed the order of injections in the constructor.

File: bundles/Billing/Service/BillingService.php (hunk 7 above)
Comment (tuj) [reply bundles/Billing/Service/BillingService.php:253]: It means that the invoice has been finished and it can not be changed.
Not the best name, I agree

File: src/Entity/AbstractEntity.php (hunk 8 above)
Comment (tuj) [reply src/Entity/AbstractEntity.php]: Removed.

**Review by rimi-itk — APPROVED** [review]
